onpremise/onpremise.go

```diff
@@ -28,55 +30,65 @@ type Engine struct {
 	isManagerInitialized          bool
 	product                       string
 	maxRetries                    int
+	lastModificationTimestamp     *time.Time
+	isFileWatcherEnabled          bool
+	isUpdateOnStartEnabled        bool
 }
 
 const (
 	defaultDataFileUrl = "https://distributor.51degrees.com/api/v2/download?Type=HashV41&Download=True&Product=V4TAC"
 )
 
+var (
+	ErrNoDataFileProvided           = errors.New("no data file provided")
+	ErrTooManyRetries               = errors.New("too many retries to pull data file")
+	ErrFileNotModified              = errors.New("data file not modified")
+	ErrLicenceKeyAndProductRequired = errors.New("licence key and product are required")
+)
+
 // run starts the engine
-func (p *Engine) run() error {
-	if len(p.dataFile) > 0 {
-		err := p.initializeManager()
+func (e *Engine) run() error {
+	if len(e.dataFile) > 0 {
+		err := e.initializeManager()
 		if err != nil {
 			return err
 		}
 	}
 
-	err := p.validateAndAppendUrlParams()
+	err := e.validateAndAppendUrlParams()
 	if err != nil {
 		return err
 	}
 
-	if p.isScheduledFilePullingEnabled {
-		go p.scheduleFilePulling()
+	if e.isScheduledFilePullingEnabled && e.isAutoUpdateEnabled {
+		go e.scheduleFilePulling()
 	}
 
 	return nil
 }
 
-func (p *Engine) initializeManager() error {
+func (e *Engine) initializeManager() error {
 
-	err := dd.InitManagerFromFile(p.manager, *p.config, "", p.dataFile)
+	err := dd.InitManagerFromFile(e.manager, *e.config, "", e.dataFile)
 	if err != nil {
 		return fmt.Errorf("failed to init manager from file: %w", err)
 	}
-	p.isManagerInitialized = true
+	e.isManagerInitialized = true
 
 	return nil
 }
 
-func (p *Engine) createDatafileIfNotExists() error {
-	_, err := os.Stat(p.dataFile)
+func (e *Engine) createDatafileIfNotExists() error {
```

Comment:
I think this function really is not needed, the data file path must always be provided, and should exist if it does not exist - then the underlying API will panic and it is normal to panic in such situation.